Bot modifications

aknarts opened the pull request:
Hi,

I did couple modifications to your bot, the most important being:

- Added handler for Build(pipeline Jobs)
- Added support for Pipelines
- Added more configuration options(Listen address, Listen port, Enabled handlers)

Furthermore I fixed some errors that happened to me while testing with the current version of gitlab(Undefined keys returned in the JSON by gitlab) and updated the documentation links

All the old functionality should be preserved with the added hooks, once people decide to update the config file the new config options should kick in

aknarts commented:
Sorry this is such a big change, did not want to create bunch of new ones, there might be some new in the future.

blha303 commented on requirements.txt:
```diff
@@ -1,3 +1,2 @@
 Flask==0.11.1
-requests==2.11.1
-discord.py
```
[phony.py](https://github.com/blha303/gitlab-discord-bridge/blob/master/phony.py) needs discord.py. phony.py isn't necessary, but someone asked if I could keep the bot online between messages, so I added it to the repo.

blha303 commented on app.py:
```diff
@@ -79,31 +132,41 @@ def post_to_discord(channel, text):
+            return make_response(post_to_discord(channelid, handlers[body["object_kind"]](body)), 200)
+    except KeyError:
+        if body["object_kind"] in handlers and "enabled_hooks" not in config:
+            return make_response(post_to_discord(channelid, handlers[body["object_kind"]](body)), 200)
```
A message should be printed to inform the user if the config needs to be updated.

aknarts replied:
Ok lets have a discussion here, you are correct that enabled_hooks are not necessary as that can be configured in gitlab, so should I just remove this entire piece of code and revert it to the original one without the enabled_hooks checks?

blha303 replied:
I'd say so, yeah. I'd like to avoid any breaking changes with anyone that tries to update the program without changing their config. To that end I'd probably remove the assert lines at the top and change `app.run` to use `config.get("listen", "0.0.0.0")` and `config.get("port", 25431)` instead. The only required lines in the config should be `secret` and `token` ideally.

aknarts replied:
Already removed the additional config and the startup does not fail when listen or port is not set up. But using config.get with the default is better than my ifs:). Going to change this once I get home.

blha303 commented on app.py:
```diff
@@ -79,31 +132,41 @@ def post_to_discord(channel, text):
-        return make_response("go away please", 403)
+        return make_response("Method Not Allowed", 405)
+    if ("host" in config and request.remote_addr != config["host"]) or request.headers.get("X-Gitlab-Token", "") != \
+            config["secret"]:
```
Unnecessary line split, please remove

aknarts replied:
Well https://www.python.org/dev/peps/pep-0008/ does recommend even shorter line length than PyCharm formating does, but yeah this one is truly unnecessary.

blha303 commented on app.py:
```diff
@@ -1,74 +1,127 @@
     """ Handle GitLab pipeline event webhook
-        https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/web_hooks/web_hooks.md#pipeline-events """
+        https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/user/project/integrations/webhooks.md#pipeline-events """
     # i don't even know what to put here
```
Thank you for implementing pipelines for me, I have no need for them myself so I'm happy to leave these functions as they are. Just remove this comment line for me please, it's not needed anymore.

aknarts replied:
Hehe missed that one :)

blha303 commented on app.py:
```diff
@@ -1,74 +1,127 @@
     return """:arrows_counterclockwise: <{object_attributes[url]}>
-**{user[name]}** created a merge request: {object_attributes[source_branch]}->{object_attributes[target_branch]} **{object_attributes[title]}**""".format(**body)
+**{user[name]}** created a merge request: {object_attributes[source_branch]}->{object_attributes[target_branch]} **{object_attributes[title]}**""".format(
+        **body)
```
Unnecessary return

blha303 commented on app.py:
```diff
@@ -1,74 +1,127 @@
+    try:
+        a = body["object_attributes"]["action"]
+    except KeyError:
+        a = "Touch"
```
Touch?

aknarts replied:
Did not know what to put there, as that is not supposed to happen but does happen when I try to test the hook, so it resultes in <Name> touched  the  issue.

blha303 commented on app.py:
```diff
@@ -1,74 +1,127 @@
     config = json.load(f)
 
+assert "secret" in config, "Gitlab secret token in config not defined"
+assert "token" in config, "Discord token in config not defined"
```
enabled_hooks check should be done up here instead of returning
In addition, why have the enabled_hooks check if gitlab has checkboxes to pick what kind of webhooks are sent to endpoints?

blha303 commented on app.py:
```diff
@@ -1,74 +1,127 @@
 #!/usr/bin/env python3
 from flask import *
 from requests import post
-from json import load, dump
```
This breaks the program. Please readd this line, otherwise the block directly below it won't work.

aknarts replied:
Strange I tested both the requirements and this one and worked...will add it back

blha303 commented:
Big changes aren't a problem, I can squash them into a single commit when I pull it.

aknarts commented:
Removed the enabled_hooks all together as they are redundant. 
Fixed illogical line breaks
Instead of Touch just return False, as that situation should not happen

A sidenote question, what License do you intend for this?:)

aknarts commented:
I think these are the last changes I am going to sneak in here, I do have a systemd file and linux install script prepared too, but I think that deserves a separate PR :)
